eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala

```diff
@@ -52,26 +51,31 @@ class PaymentInitiator(nodeParams: NodeParams, outgoingPaymentFactory: PaymentIn
         // Immediately return the paymentId
         sender ! paymentId
       }
-      val paymentCfg = SendPaymentConfig(paymentId, paymentId, r.externalId, r.paymentHash, r.recipientAmount, r.recipientNodeId, Upstream.Local(paymentId), r.paymentRequest, storeInDb = true, publishEvent = true, Nil)
+      val paymentCfg = SendPaymentConfig(paymentId, paymentId, r.externalId, r.paymentHash, r.recipientAmount, r.recipientNodeId, Upstream.Local(paymentId), Some(r.paymentRequest), storeInDb = true, publishEvent = true, Nil)
       val finalExpiry = r.finalExpiry(nodeParams.currentBlockHeight)
-      r.paymentRequest match {
-        case Some(invoice) if !invoice.features.areSupported(nodeParams) =>
-          sender ! PaymentFailed(paymentId, r.paymentHash, LocalFailure(Nil, UnsupportedFeatures(invoice.features.features)) :: Nil)
-        case Some(invoice) if invoice.features.allowMultiPart && nodeParams.features.hasFeature(BasicMultiPartPayment) =>
-          invoice.paymentSecret match {
-            case Some(paymentSecret) =>
-              val fsm = outgoingPaymentFactory.spawnOutgoingMultiPartPayment(context, paymentCfg)
-              fsm ! SendMultiPartPayment(sender, paymentSecret, r.recipientNodeId, r.recipientAmount, finalExpiry, r.maxAttempts, r.assistedRoutes, r.routeParams, userCustomTlvs = r.userCustomTlvs)
-            case None =>
-              sender ! PaymentFailed(paymentId, r.paymentHash, LocalFailure(Nil, PaymentSecretMissing) :: Nil)
-          }
-        case _ =>
-          val paymentSecret = r.paymentRequest.flatMap(_.paymentSecret)
+      r.paymentRequest.paymentSecret match {
+        case _ if !r.paymentRequest.features.areSupported(nodeParams) =>
+          sender ! PaymentFailed(paymentId, r.paymentHash, LocalFailure(Nil, UnsupportedFeatures(r.paymentRequest.features.features)) :: Nil)
+        case None =>
+          sender ! PaymentFailed(paymentId, r.paymentHash, LocalFailure(Nil, PaymentSecretMissing) :: Nil)
+        case Some(paymentSecret) if r.paymentRequest.features.allowMultiPart && nodeParams.features.hasFeature(BasicMultiPartPayment) =>
+          val fsm = outgoingPaymentFactory.spawnOutgoingMultiPartPayment(context, paymentCfg)
+          fsm ! SendMultiPartPayment(sender, paymentSecret, r.recipientNodeId, r.recipientAmount, finalExpiry, r.maxAttempts, r.assistedRoutes, r.routeParams, userCustomTlvs = r.userCustomTlvs)
+        case Some(paymentSecret) =>
           val finalPayload = Onion.createSinglePartPayload(r.recipientAmount, finalExpiry, paymentSecret, r.userCustomTlvs)
           val fsm = outgoingPaymentFactory.spawnOutgoingPayment(context, paymentCfg)
           fsm ! SendPayment(sender, r.recipientNodeId, finalPayload, r.maxAttempts, r.assistedRoutes, r.routeParams)
       }
 
+    case r: SendSpontaneousPaymentRequest =>
```

Comment:
True, I've never been a big fan of the names of these types anyway, but wanted to avoid clashing with the `PaymentLifecycle` and `MultiPartPaymentLifecycle` ones. When we move those to akka-typed, we will put them in their own object which will let us harmonize this (these payment actors should be quite easy to migrate to akka-typed when we have some free time).

Reply:
I took this opportunity to do it right now in https://github.com/ACINQ/eclair/pull/1840/commits/64e0e365477e98b3e5dde27bf349bc90a6ef34bc

We need to prefix with `PaymentLifecycle` or `PaymentInitiator` when it's ambiguous, but that only happens in tests and in the `PaymentInitiator`, so it's fine.

Reply:
👌